### PR TITLE
Add repetition_penalty aligned with huggingface

### PR DESCRIPTION
**Pull request by Abraham-Xu**
In huggingface/transformers generate method: 
https://github.com/huggingface/transformers/blob/ae320fa53f74cc4dfa0e4fc3c95b6129a86b0512/src/transformers/generation/utils.py#L1295
1. it prepares a logits_processor from generation_config
https://github.com/huggingface/transformers/blob/ae320fa53f74cc4dfa0e4fc3c95b6129a86b0512/src/transformers/generation/utils.py#L1540
2. it does logits pre-process by logits_processor in greedy_search/sample/beam_search
https://github.com/huggingface/transformers/blob/ae320fa53f74cc4dfa0e4fc3c95b6129a86b0512/src/transformers/generation/utils.py#L2457

Specificaly, **repetition penalty** is a frequently used pre-process method in logits_processor. It prevents repetitions through a penalty. This penalized sampling works by discounting the scores of previously generated tokens. (https://arxiv.org/pdf/1909.05858.pdf)

I introduced the the repetition penalty pre-process to sampler.py aligned with huggingface implementation:
https://github.com/huggingface/transformers/blob/ae320fa53f74cc4dfa0e4fc3c95b6129a86b0512/src/transformers/generation/logits_process.py#L328

### CODE REVIEW

**Review by Scagin — APPROVED** [review]
I am waiting for this, I need some method to make hf and vllm compat when I use penalty arguments

**Comment by Abraham-Xu** [comment]
I have tested the repetition_penalty under greedy search mode by Llama 7b and Baichuan models. The results are all aligned with huggingface transformers.

File: vllm/sampling_params.py [hunk 1]
```diff
@@ -94,6 +99,9 @@ def _verify_args(self) -> None:
         if not -2.0 <= self.frequency_penalty <= 2.0:
             raise ValueError("frequency_penalty must be in [-2, 2], got "
                              f"{self.frequency_penalty}.")
+        if self.repetition_penalty <= 0.0:
+            raise ValueError("repetition_penalty must be a strictly positive "
+                             f"float, got {self.repetition_penalty}.")
```
Comment (pfldy2850) [inline vllm/sampling_params.py:104]:
```suggestion
        if self.repetition_penalty <= 0.0:
            raise ValueError("repetition_penalty must be a strictly positive "
                             f"float, got {self.repetition_penalty}.")
        if self.repetition_penalty != 1.0 and (
                abs(self.frequency_penalty) > _SAMPLING_EPS
                or abs(self.presence_penalty) > _SAMPLING_EPS):
            raise ValueError(
                f"repetition_penalty cannot be used with "
                f"frequency_penalty and presence_penalty."
                f"got repetition_penalty={self.repetition_penalty}, "
                f"frequency_penalty={self.frequency_penalty}, "
                f"presence_penalty={self.presence_penalty}")
```


It seems that repetition penalty and other penalties cannot be used together, how about exception handling like above?

**Comment by markluofd** [comment]
this pr solved my problem, for my llama2 model, repetition_penalty is necessary, otherwise the result is incorrect

**Comment by TheBloke** [comment]
Great! Looking forward to testing this as my perception is that vLLM is very poor with repetition currently, and that the existing repetition penalty params (freqency_penalty etc) do little or nothing.

Although part of that problem is that there's no per-request seed, something we also really need.

**Comment by jeffchy** [comment]
Beautiful！

**Comment by BaiMoHan** [comment]
Very nice

**Review by ghost — APPROVED** [review]

**Comment by BaiMoHan** [comment]
Hey, what' wrong?There is no repetition_penalty  attr in sampling_params when I use v0.1.7.

**Comment by BaiMoHan** [comment]
> Hey, what' wrong?There is no repetition_penalty attr in sampling_params.py when I use v0.1.7. I'm a bit impatient.🥺🥺🥺

File: vllm/model_executor/layers/sampler.py [hunk 2]
```diff
@@ -162,30 +170,61 @@ def _apply_penalties(
         indices.append(i)
 
     # Return early if all sequences have zero penalties.
```
Comment (leshanbog) [inline vllm/model_executor/layers/sampler.py:172]: this comment is misleading now?

File: vllm/model_executor/layers/sampler.py [hunk 3]
```diff
@@ -162,30 +170,61 @@ def _apply_penalties(
...
+        presence_penalties = torch.tensor(presence_penalties,
+                                          dtype=logits.dtype,
+                                          device=logits.device)
+        # We follow the definition in OpenAI API.
+        # Refer to
+        # https://platform.openai.com/docs/api-reference/parameter-details
+        logits[indices] -= frequency_penalties.unsqueeze(dim=1) * bin_counts
+        presence_mask = (bin_counts > 0.0).to(dtype=logits.dtype)
+        logits[indices] -= presence_penalties.unsqueeze(dim=1) * presence_mask
+    else:
```
Comment (leshanbog) [inline vllm/model_executor/layers/sampler.py:196]: why `else`? will presence_penalty and `repetition_penalty` work together?

**Comment by heshuguo** [comment]
@WoosukKwon please check if the fix can be merged.

**Comment by WoosukKwon** [comment]
@Abraham-Xu @heshuguo This PR is a bit in conflict with #1048, which tries to refactor the sampler code. @zhuohan123 Can we merge this PR before yours?

**Comment by WoosukKwon** [comment]
Hi @Abraham-Xu, sorry for the late response.

Could you update the PR? There are some merge conflicts because we refactored sampler in #1048.

**Comment by jessiewiswjc** [comment]
I want to know if `presence_penalty/frequency_penalty` in vllm can work with `repetition_penalty` in hf at the same time?

**Comment by zhuohan123** [comment]
Supported with #1424. But again, thank you for your contribution! Let us know if there is any other issue.

**Comment by Abraham-Xu** [comment]
> Supported with #1424. But again, thank you for your contribution! Let us know if there is any other issue.

Sorry for late reply, busy with working and other things recently. It is great to support repetition penalty in any way.
